11.0 auditlog extension

gfcapalbo opened the pull request:
@thomaspaulb first alpha

gfcapalbo committed:
[ADD] pre-review version

thomaspaulb commented on auditlog/views/auditlog_view.xml:
```diff
@@ -193,7 +193,7 @@
         <field name="res_model">auditlog.log</field>
         <field name="view_type">form</field>
-        <field name="search_view_id" ref="view_auditlog_log_search"/>
+        <field name="search_view_id" ref="auditlog.view_auditlog_log_search"/>
```
This should not be needed?

gfcapalbo replied:
was removed in latest push

thomaspaulb replied:
It seems to be back

thomaspaulb replied:
@gfcapalbo Explanation for this?

thomaspaulb commented on auditlog_security/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+
+{
+    'name': "Audit Log Permissions",
+    'version': "11.0.1.0.1",
```
We start with `11.0.1.0.0`, and I think a better name would be `auditlog_fields`.

But leave this for the moment, let's think about the functionality first. If there are no downsides to using it (eg. it won't become slower or more buggy), then I suggest we just move it directly into the main `auditlog` module.

thomaspaulb commented on auditlog_security/data/auditlog_rule.xml:
```diff
@@ -0,0 +1,9 @@
+            <field name="model_id" ref="base.model_res_partner"/>
+        </record>
+
+</odoo>
```
This should be demo data.

gfcapalbo replied:
it was conceived as normal data, but as i completed it i realized it is not needed. user can just create rule. 
goes in /demo

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,55 @@
+        'res.groups', 
+        string='Groups',
+        help="""Groups that will be allowed to see the logged fields, if left empty 
+                all groups will be allowed (global rule creation)"""
```
I see what you mean, but what's the use case? Global rules are [restrictive](https://sfithub.com/sunflowerit/odoo-docs/-/wikis/Record-security), where group rules are additive. So let's say we do support this, then yes, all groups will be allowed to see this, but no more than that. It would probably even be destructive because if there's a global rule saying that you can see res.partner rules, then that automatically means no-one can see hr.employee rules anymore.

So I'd say that this should always be filled with a group, and if it's not filled we default to creating the group for the "internal user".

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,55 @@
+
+    field_ids = fields.Many2many(
+            'ir.model.fields',
+            required=True,)
```
It can't be required always, because we need to keep supporting pre-existing rules that are just there but don't (need to) have our new cool fields functionality. So I say we remove the requiredness here, and make it required only in the form view and only when a checkbox is enabled saying that for this rule, we want to use the "Publicly visible log lines" subfunctionality.

gfcapalbo replied:
I hadn't thought about that , you are correct. we will still have "normal" auditlog.rule.

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,55 @@
+            'perm_read': True,
+            'domain_force': domain_force,
+            'auditlog_id': self.id,
+        })
```
Refactor this part away into a `_prepare_rule_values` function, which can be overridden by submodules if needed.

thomaspaulb commented on auditlog_security/models/ir_rule.py:
```diff
@@ -0,0 +1,23 @@
+        # if this.auditlog_id and not self.context('auditlog_write'):
+        #     raise exceptions.ValidationError(
+        #             _("""This rule was automatically generated from the auditlog 
+        #             interface, please use that to edit/delete it """))
```
I say we allow editing, but not deleting. Editing can come in handy in case of, well, Idunno. Shenanigans.

gfcapalbo replied:
Okay. But we will forbid changing the autolog_id field.  
that is the connection with our rule. it should not be manually changed/lost.

My concern was that there is a difference between the autolog.rule and the ir.rule , that may confuse the user.
but editing is feasable it we accept that

thomaspaulb commented on auditlog_security/tests/test_auditlog.py:
```diff
@@ -0,0 +1,121 @@
+
+    def tearDown(self):
+        self.groups_rule.unlink()
+        super(TestAuditlogFast, self).tearDown()
```
Remove tests, they're not ours

gfcapalbo replied:
tests were removed.

gfcapalbo committed:
[ADD] review fixes

gfcapalbo commented:
@tom all review fixes done and tested , found more needed features:

 [ x ] version change
 [ x ] removed demo data , not needed
 [ x ] groups_id are not required but if  fields_ids is set , we add the default res_users
 [ x ] checkbox for when we select no "field_ids", to signify  we have created a rule for all fields   ( readonly compute, just for clarity)
 [ x ] onchange: if model_id is changed , we MUST wipe out currently selected field_ids  ( very important)
 [ x ] create did not work in creating rule , now does
 [ x ]  refactored as requested
 [ x ] domain works even when group is selected , but no fields.

gfcapalbo commented:
ADDED : 
now we manage the case of:
    auditlog.rule with permissions, becoming a "normal" rule,  we now delete the associated rules.

gfcapalbo commented:
Added:
       Now "subscribe" is overwritten to create the action we want. automatic generation of action , works on any model.
       removed hardcoded action.

gfcapalbo commented:
@tom Needed to add the automatic action as a feature:
      The "subscribe"  in audit log would create it's own action, after we made the /DATA /DEMO, having the hardcoded res.partner action made no more sense.
It was quick and easy to do.
We can extend it however we want. 
"unsubscribe" just deletes the action.

gfcapalbo commented:
removed the "all_fields" boolean , works perfectly without it.

gfcapalbo commented:
@thomaspaulb

gfcapalbo commented:
@thomaspaulb reminder for final approval.

thomaspaulb commented on auditlog_security/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Audit Log Permissions",
```
I don't think this really describes the functionality of the module... I'm thinking more of 'Audit Log - User Access' or something. Or even User Permissions. What do you think ?

gfcapalbo replied:
I think "audit log - user permissions"

thomaspaulb commented on auditlog_security/models/auditlog_autovacuum.py:
```diff
@@ -0,0 +1,14 @@
+    _inherit = "auditlog.autovacuum"
+
+    @api.model
+    def autovaccum(self, days):
```
There's a typo here. Puccini, cappuccino... understandable, but no.

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,118 @@
+            ("name", "not in", FIELDS_BLACKLIST),
+        ]
+
+    def unlink(self):
```
For Odoo 11.0 we still need explicit `@api.multi`.

thomaspaulb commented on auditlog_security/models/ir_rule.py:
```diff
@@ -0,0 +1,48 @@
+        "auditlog.rule",
+        required=False,
+        help="Auditlog Rule that generated this ir.rule",
+    )
```
We should add `index=True` here, because elsewhere you're using it to search on it.

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,118 @@
+
+    def get_linked_rules(self):
+        # return with context key so that deletion will not be forbidden
+        return self.with_context(auditlog_write=True).env["ir.rule"].search(
```
It doesn't feel right to me to include this context here... rather directly at the place where you are calling the `unlink` or `write`.

thomaspaulb commented on auditlog_security/models/ir_rule.py:
```diff
@@ -0,0 +1,48 @@
+
+    auditlog_id = fields.Many2one(
+        "auditlog.rule",
+        required=False,
```
Should we add here an `ondelete='cascade'`? Then the `ir.rule` is deleted whenever the auditlog rule is deleted, which I think is what we want, and we don't need to bother with custom `unlink` overrides.

thomaspaulb replied:
@gfcapalbo Rationale for not adding it?

gjotten replied:
@gfcapalbo this right here looks like it needs some attention

gfcapalbo replied:
pre-email cleanup comment. had missed.

thomaspaulb commented on auditlog_security/readme/USAGE.rst:
```diff
@@ -0,0 +1,10 @@
+The rule is now extended with a new field permission_ids, that tells us wich groups will
+be allowed to read the lines produced by this rule.
+If permission_ids is left empty, the default will be: 
+"auditlog lines visible only by administrator"
```
This is technically not true, it's "by anyone who is in Settings group, which is the default for the auditlog module"

https://github.com/sunflowerit/server-tools/blob/11.0/auditlog/security/ir.model.access.csv

thomaspaulb commented on auditlog_security/security/ir.model.access.csv:
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_auditlog_log_line_user,auditlog_log_line_user,model_auditlog_log_line,base.group_user,1,0,0,0
+access_auditlog_log_user,auditlog_log_user,model_auditlog_log,base.group_user,1,0,0,0
```
I understand the first one, but do we need the second one? I thought we were only giving access to lines.

gfcapalbo replied:
This was from when we showed the rule and filtered line.
Standard auditlog still creates the"Logs" action, so with line 2, instead of an error, we see the read-only log, with ONLY the lines user can see, if any.
Looks better than an error on action execution IMHO.

thomaspaulb replied:
@gfcapalbo Yeaaaaa true but it could also give users access to something they shouldnt see. Shouldnt the real solution be to have the 'View Logs' button only visible to admins?

thomaspaulb commented on auditlog_security/views/auditlog_view.xml:
```diff
@@ -0,0 +1,114 @@
+    
+
+
+    <!-- TO BE REMOVED, KEPT FOR TESTING PURPOSES
```
Ah to be honest this is useful, maybe you can just keep it but instead place it in `auditlog` module directly, but make it available to `base.group_no_one` only, so that it's only visible in Debug mode.

thomaspaulb commented on auditlog_security/views/auditlog_view.xml:
```diff
@@ -0,0 +1,114 @@
+
+
+    <menuitem id="menu_audit_log_lines" name="Log Lines"
+        parent="menu_audit_log_line" action="action_auditlog_log_line_tree"/>
```
The correct location in that case would be below this one:

https://github.com/sunflowerit/server-tools/blob/11.0/auditlog/views/auditlog_view.xml#L200

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,118 @@
+
+    def unlink(self):
+        # if we delete auditlog rule, corresponding ir.rules are removed
+        # TODO PROPOSAL: a warning here with detailed information?
```
Nah I don't think that's needed, in fact I think we should just silently cascade-delete them away. But shouldn't that be automatic? See also my comment below about `ondelete=cascade`

gjotten replied:
@gfcapalbo this is unresolved as well

_edit:_ though outdated?

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,118 @@
+        return [
+            ("model_id", "=", self.env.ref("base.model_res_partner").id),
+            ("name", "not in", FIELDS_BLACKLIST),
+        ]
```
This is currently not called so maybe you can comment it out so it's clear in the review that it's a draft thing

thomaspaulb commented on auditlog_security/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    "author": "Therp B.V.,Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "website": "https://github.com/OCA/server-tools/",
+    "category": "Tools",
```
Then as a "description" subtitle we can use "Allow regular users to view Audit log lines via the form view of the relevant model"

thomaspaulb commented on auditlog_security/views/auditlog_view.xml:
```diff
@@ -0,0 +1,114 @@
+        <field name="inherit_id" ref="auditlog.view_auditlog_rule_form"/>                         
+        <field name="arch" type="xml">                                                  
+            <xpath expr="//group[last()]" position="after">          
+                <group>
```
It's currently not so obvious to the user what these fields and groups do. Maybe we can add here a div with a small text such as:

```
Add fields here to make any changes to them (audit log lines) visible to members of the selected groups.
```

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,118 @@
+        "res.groups",
+        help="""Groups that will be allowed to see the logged fields, if left empty 
+                default will be all users with a login""",
+    )
```
Actually if we do it like this, it's not possible to for example say: for this particular audit log rule eg on `hr.employee`, we will give access to HR managers to see changes to `job_id`, and to HR recruiters only to `contract_date_end` or something. We can just give all listed groups access to all the listed fields, and not differentiate.

Am I right or seeing something wrong? If it is so, then that's something we should change, and IMO it would also make it look clearer - we make a `<group>` called "User Access" which shows a list view that leads to a new model `auditlog.line.access.rule`, and that model would have many "fields_ids / group_ids" combinations, that would each be linked with a particular `ir.rule` that makes that particular access possible.

gfcapalbo replied:
I had made the extra model in my first iteration, still have that branch.
But in this case, in the example you made we can just create:
            1 auditlog rule with   field_ids: ;'job_id'     and group_id: HR_manager
            then another auditlog rule with : field_ids: 'contract_date_end' and group_id:  'HR_recruiters'

this would create two seperate rules for the two different groups. It;s simpler and creates less complex rules.
I think it's clearer than putting everything in a separate model, after making the "extra model" option I saw it was redundant, it seems cleaner to put everything in auditlog_rule.
But Now i see I made a terrible mistake, there is a unique constraint on auditlog.rule. We need the extra model. 
On it.

gfcapalbo replied:
Solved and now we have maximum flexibility for N gorup-field relations.

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,118 @@
+        }
+
+    @api.multi
+    def subscribe(self):
```
Hey.... we should call `super()` here, and leave the creation of the `View logs` action of the parent module intact. Right? Removing that functionality is not "inherited module boy scout behaviour".

ON TOP OF THAT we can add our "View log lines" action, which is the added feature of our particular module.

gfcapalbo replied:
Ah! yes , the original is written in a way to make it un-inheritable. 
It creates  "old" action
and does not return the created action, just true.

In order to preserve inheritance i should 

               call super
               find the action it created and modify the domain with what we want.
             
will do that ( we just have to modify all actions of the rule after calling super) and we preserve inheritance chain.

thomaspaulb replied:
@gfcapalbo I'm not sure if I agree with this for OCA. For customer it's fine but this is destroying the 'View Logs' functionality which people may still want to have. I think we should make the "View Logs" available for admins, and the "View Log Lines" for any user that is member of a group that is allowed to see. So an admin could see both buttons, but a user only one.

gfcapalbo replied:
"view logs" intact . additional "view log lines" added.

thomaspaulb commented on auditlog_security/models/auditlog_rule.py:
```diff
@@ -0,0 +1,118 @@
+                'res_model': 'auditlog.log.line',
+                'src_model': rule.model_id.model,
+                'binding_model_id': rule.model_id.id,
+                'domain': domain,
```
I think we should fill `groups_id` here with the result of `rule.mapped('groups_ids')`.

Then we should also make the `fields_ids` and `groups_ids` readonly once the rule is subscribed, to avoid us having to write functionality to keep the both in sync when groups are added or removed.

thomaspaulb commented:
Actually what i meant was to create two actions, view logs, and view log lines. So you can leave the "view logs" action as is. They do different things

May 2, 2022 10:38:52 Giovanni Capalbo (Therp) ***@***.***>:

> ***@***.**** commented on this pull request.
> 
> ----------------------------------------
> 
> In auditlog_security/models/auditlog_rule.py[https://github.com/sunflowerit/server-tools/pull/4#discussion_r862680512]:
> 
>> +    def _prepare_rule_values(self):
> +        domain_force = "[" + " ('log_id.model_id' , '=', %s)," % (self.model_id.id)
> +        if self.field_ids:
> +            domain_force += "('field_id', 'in',  %s)" % (self.field_ids.ids)
> +        domain_force += "]"
> +        return {
> +            "name": "auditlog_extended_%s" % self.id,
> +            "model_id": self.env.ref("auditlog.model_auditlog_log_line").id,
> +            "groups": [(6, 0, self.group_ids.ids)],
> +            "perm_read": True,
> +            "domain_force": domain_force,
> +            "auditlog_id": self.id,
> +        }
> +
> +    @api.multi
> +    def subscribe(self):
> 
> Ah! yes , the original is written in a way to make it un-inheritable.
> It creates "old" action
> and does not return the created action, just true.
> 
> In order to preserve inheritance i should
> 
>            call super
>            find the action it created and modify the domain with what we want.
> 
> will do that ( we just have to modify all actions of the rule after calling super) and we preserve inheritance chain.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/sunflowerit/server-tools/pull/4#discussion_r862680512], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AALF75FFONAXQZMTA2HBQLLVH6IBNANCNFSM5TGIP5TQ].
> You are receiving this because you were mentioned.
>  [data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEgAAABICAYAAABV7bNHAAAAAXNSR0IArs4c6QAAAARzQklUCAgICHwIZIgAAAArSURBVHic7cEBDQAAAMKg909tDjegAAAAAAAAAAAAAAAAAAAAAAAAAAA+DFFIAAEctgHwAAAAAElFTkSuQmCC###24x24:true###][Tracking image][https://github.com/notifications/beacon/AALF75FVTPDZERRHC6C2Z4DVH6IBNA5CNFSM5TGIP5T2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOHETLOPQ.gif]
>

gfcapalbo commented:
@thomaspaulb added a video showing it working on issue tracker, and added some small cleanups this morning.
Ready for review, all errors/comments addressed except that 1 on security, where I explain the sense of it.

thomaspaulb commented on auditlog_security/models/auditlog_line_access_rule.py:
```diff
@@ -0,0 +1,113 @@
+class AuditlogLineAccessRule(models.Model):
+    _name = "auditlog.line.access.rule"
+
+    name = fields.char()
```
@gfcapalbo This should be capitalized, it's now breaking the test servers.

gfcapalbo commented:
@thomaspaulb 
- [ x ] fixed the breaking typo
- [ x ] removed useless XMLID change in "auditlog" module
- [ x ] subscribe does not delete original action, just adds new one.
- [ x ] rebuilt README

thomaspaulb commented on auditlog_security/models/__init__.py:
```diff
@@ -0,0 +1,6 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from . import auditlog_rule
+from . import auditlog_line_access_rule
```
Upon installing the module I saw: `Field ir.rule.auditlog_line_access_rule_id with unknown comodel_name 'auditlog.access.line.rule'` - maybe that goes away when the order here should be switched?

gfcapalbo replied:
not sure, but makes sense. switched.

thomaspaulb replied:
Turned out to be a typo, see my other comment.

thomaspaulb commented on auditlog_security/views/auditlog_view.xml:
```diff
@@ -0,0 +1,119 @@
+                                </field>
+                                    <field name="field_ids"
+                                        domain="[('model_id', '=' , model_id),('name', 'not in', ['id', 'create_uid', 'create_date', 'write_uid', 'write_date','display_name', '__last_update']) ]"
+                                        attrs="{'readonly': [('state', '=', 'subscribed')]}"
```
These two references to `state` field are giving issues:

```
Field 'state' used in attributes must be present in view but is missing:
 - 'state' in attrs="{'readonly': [('state', '=', 'subscribed')]}"
 - 'state' in attrs="{'readonly': [('state', '=', 'subscribed')]}"
```

Not seeing `state` field in the model of `auditlog.rule` anymore, it was moved to the request object, so I guess this can be removed?

gfcapalbo replied:
double checking.

gfcapalbo replied:
State is in the model auditlog rule, this attrs is in a embedded view , and it refers to model: "auditlog.line.access.rule",
wich in fact does not have 'state'. 


i will add 'state' as a reference field to it's connected rule
auditlog.line.access.rule  is connected to auditlog.rule via M2O ,  so I will impose auditlog.line.access.rule    the  to have the same state of  it's associated rule , and then add it to view.

gfcapalbo replied:
@thomaspaulb  solved by https://github.com/sunflowerit/server-tools/pull/4/commits/4cabfa5aae4a6bb4319e1bbb63efffe1f6bf0b11 
Untested, I can test after I'm done with a current debug session on another project

thomaspaulb commented on auditlog_security/models/ir_rule.py:
```diff
@@ -0,0 +1,51 @@
+    _inherit = "ir.rule"
+
+    auditlog_line_access_rule_id = fields.Many2one(
+        "auditlog.access.line.rule",
```
This should be `auditlog.line.access.rule`.

thomaspaulb commented on auditlog_security/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+        "contacts",
+    ],
+    "data": [
+        "views/auditlog_view.xml",
```
Since we're not including `ir.model.access.csv` here, we're not loading access rules and so we get problems for non-admin users.

gfcapalbo commented:
@thomaspaulb i don;t understand how was the CSV import removed , testing for regressions.

gfcapalbo commented:
@thomaspaulb 
Re-checked all open issues. rebased, only 2 commits now.

thomaspaulb commented:
Closed in favour of https://github.com/Therp/server-tools/pull/1

(I can push there too)
